oak/server/rust/oak_runtime/src/node/wasm.rs

```diff
@@ -0,0 +1,806 @@
+const WAIT_ON_CHANNELS: usize = 6;
+
+type AbiHandle = u64;
+type AbiPointer = u32;
+type AbiPointerOffset = u32;
+
+const ABI_U32: ValueType = ValueType::I32;
+const ABI_U64: ValueType = ValueType::I64;
+const ABI_USIZE: ValueType = ValueType::I32;
+
+/// `WasmInterface` holds runtime values for a particular execution instance of Wasm. This includes
+/// the host ABI function mapping between the runtime and a Wasm instance, and the handle mapping
+/// between the instance and the runtime `ChannelRef`s.
+///
+/// Any handle from `readers` or `writers` is unique to all handles in both `readers` and
+/// `writers`.
+struct WasmInterface {
+    /// A configuration and thread specific name for pretty printing.
+    pretty_name: String,
+
+    /// A reference to the `Runtime` is needed to execute the host ABI functions.
+    runtime: RuntimeRef,
+
+    /// Reader channel mappings to unique u64 handles
+    readers: HashMap<AbiHandle, ChannelReader>,
+    /// Writer channel mappings to unique u64 handles
+    writers: HashMap<AbiHandle, ChannelWriter>,
+
+    /// A reference to the memory used by the `wasmi` interpreter. Host ABI functions using Wasm
+    /// relative addresses will perform reads/writes against this reference.
+    memory: Option<wasmi::MemoryRef>,
+}
+
+impl WasmInterface {
+    /// Generate a randomized handle. Handles are random to prevent accidental dependency on
+    /// particular values or sequence. See https://github.com/project-oak/oak/pull/347
+    fn allocate_new_handle(&mut self, channel: ChannelEither) -> AbiHandle {
+        loop {
+            let handle = rand::thread_rng().next_u64();
+
+            if self.readers.get(&handle).is_some() || self.writers.get(&handle).is_some() {
+                continue;
+            }
+
+            match channel {
+                ChannelEither::Reader(channel) => {
+                    self.readers.insert(handle, channel);
+                }
+                ChannelEither::Writer(channel) => {
+                    self.writers.insert(handle, channel);
+                }
+            }
+
+            return handle;
+        }
+    }
+
+    /// Helper function to get memory.
+    fn get_memory(&self) -> &wasmi::MemoryRef {
+        self.memory
+            .as_ref()
+            .expect("WasmInterface memory not attached!?")
+    }
+
+    /// Make sure a given address range falls within the currently allocated range of memory.
+    fn validate_ptr(&self, addr: AbiPointer, offset: AbiPointerOffset) -> Result<(), OakStatus> {
+        let byte_size: wasmi::memory_units::Bytes = self.get_memory().current_size().into();
+
+        if byte_size < wasmi::memory_units::Bytes((addr as usize) + (offset as usize)) {
+            return Err(OakStatus::ERR_INVALID_ARGS);
+        }
+
+        Ok(())
+    }
+
+    /// Creates a new `WasmInterface` structure.
+    pub fn new(
+        pretty_name: String,
+        runtime: RuntimeRef,
+        initial_handle: ChannelReader,
```

Comment:
super-nit: I'd avoid naming things "_handle" when they're references inside the runtime; I think it makes things clearer to only use "handle" when dealing with an integer value that's within the numbering space of a particular Node.